javatests/arcs/android/storage/service/StorageChannelImplTest.kt

```diff
@@ -0,0 +1,237 @@
+import arcs.android.storage.toProto
+import arcs.core.crdt.CrdtData
+import arcs.core.crdt.CrdtOperation
+import arcs.core.data.CountType
+import arcs.core.storage.ProxyCallback
+import arcs.core.storage.ProxyMessage
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StoreOptions
+import arcs.core.storage.UntypedActiveStore
+import arcs.core.storage.keys.RamDiskStorageKey
+import arcs.core.storage.testutil.NoopActiveStore
+import com.google.common.truth.Truth.assertThat
+import com.nhaarman.mockitokotlin2.eq
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.verify
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class StorageChannelImplTest {
+  private val DUMMY_MESSAGE = ProxyMessage.SyncRequest<CrdtData, CrdtOperation, Any?>(0)
+
+  private lateinit var storageKey: StorageKey
+  private lateinit var messageCallback: IMessageCallback
+  private lateinit var resultCallback: FakeResultCallback
+  private lateinit var onProxyMessageCallback: suspend (StorageKey, ProxyMessage<*, *, *>) -> Unit
+
+  @Before
+  fun setUp() {
+    storageKey = RamDiskStorageKey("myCount")
+    messageCallback = mock {}
+    resultCallback = FakeResultCallback()
+    onProxyMessageCallback = { storageKey: StorageKey, proxyMessage: ProxyMessage<*, *, *> -> }
+  }
+
+  @Test
+  fun proxiesMessagesFromStore() = runBlockingTest {
+    var proxyCallback: ProxyCallback<CrdtData, CrdtOperation, Any?>? = null
+    val store = object : NoopActiveStore(
+      StoreOptions(
+        storageKey,
+        CountType()
+      )
+    ) {
+      override suspend fun on(callback: ProxyCallback<CrdtData, CrdtOperation, Any?>): Int {
+        proxyCallback = callback
+        return 123
+      }
+    }
+
+    // Create channel and check it registers a listener.
+    createChannel(scope = this, store = store)
+    assertThat(proxyCallback).isNotNull()
+
+    // Check channel proxies message back.
+    proxyCallback!!.invoke(DUMMY_MESSAGE)
+    val proto = StorageServiceMessageProto.newBuilder()
+      .setProxyMessage(DUMMY_MESSAGE.toProto())
+      .build()
+    verify(messageCallback).onMessage(eq(proto.toByteArray()))
+  }
+
+  @Test
+  fun idle_waitsForStoreIdle() = runBlockingTest {
+    val job = Job()
+    val store = object : NoopActiveStore(
+      StoreOptions(
+        storageKey,
+        CountType()
+      )
+    ) {
+      override suspend fun idle() {
+        assertThat(resultCallback.hasBeenCalled).isFalse()
+        job.complete()
+      }
+    }
+
+    val channel = createChannel(scope = this, store = store)
+
+    channel.idle(1000, resultCallback)
+    job.join()
+
+    val result = resultCallback.waitForResult()
+    assertThat(result).isNull()
+  }
+
+  @Test
+  fun idle_propagatesExceptions() = runBlockingTest {
```

Comment:
I feel like some of these could be tests for the base class itself, and then we wouldn't need to duplicate them for both subclasses. e.g. all of these could be tests for the base class: `idle_propagatesExceptions`, `idle_whenChannelIsClosed_returnsError`, `idle_waitsForStoreIdle`. In the base class test `idle_waitsForStoreIdle` would just check that the `.storeIdle` method was called, but we could repeat that one in the subclass tests to actually check that the `ActiveStore`/`DirectStoreMuxer` was called (essentially, testing that the subclass overrides the `storeIdle` method properly).

What do you think?

Reply:
Yes I think that's a good idea. I created `BaseStorageChannelTest`, and I included `close_unregistersListeners` and `close_whenChannelIsClosed_returnsError` in addition to the test targets you listed above. Note that `close_unregisterListeners` is repeated in the subclass tests for the same reason `idle_waitsForStoreIdle` is repeated in the subclass tests.